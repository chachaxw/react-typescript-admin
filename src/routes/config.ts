export interface RouteConfig {
  name: string;
  icon: string;
  path: string;
  auth?: string;
  query?: string;
  hideInMenu?: boolean;
  component: string;
  children?: RouteConfig[];
}

export default {
  app: [
    {
      name: '控制台',
      icon: 'dashboard',
      path: '/app/dashboard',
      component: 'Dashboard',
    },
    {
      name: '订单中心',
      icon: 'user',
      path: '/app/order',
      component: 'OrderCenter',
    },
    {
      name: '地图中心',
      icon: 'environment',
      path: '/app/map',
      component: 'MapCenter',
    },
    {
      name: '设置中心',
      icon: 'setting',
      path: '/app/setting',
      component: 'SettingCenter',
    },
    {
      name: '用户中心',
      icon: 'user',
      path: '/app/user',
      component: 'UserCenter',
    },
  ],
  user: [],
};
