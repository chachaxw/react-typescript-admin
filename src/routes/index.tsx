import { Redirect, Route, RouteComponentProps, Switch } from 'dva/router';
import React, { Component } from 'react';
import DocumentTitle from 'react-document-title';
import routesConfig, { RouteConfig } from './config';

import queryString from 'query-string';
import { Auth } from '../models/global';
import AllPages from '../pages';
import NoPermission from '../pages/Exception/403';
import NotFound from '../pages/Exception/404';
import ServerError from '../pages/Exception/500';

interface InternalProps {
  app: any;
  auth: Auth;
}

export default class AppRoutes extends Component<InternalProps> {

  public requireAuth(permission: string, component: any) {
    const { auth } = this.props;
    const { permissions } = auth;

    if (!permissions || !permissions.includes(permission)) {
      return <Redirect to="/404" />;
    }

    return component;
  }

  public requireLogin(component: any, permission?: string) {
    const { auth } = this.props;
    const { isAuthenticated, permissions } = auth;

    if (!isAuthenticated || !permissions) { // 判断是否登录
      return <Redirect to={'/login'} />;
    }

    return permission ? this.requireAuth(permission, component) : component;
  }

  public render() {
    const { app } = this.props;

    return (
      <Switch>
        {Object.keys(routesConfig).map((key: string) =>
          routesConfig[key].map((config: RouteConfig, index: number) => {
            const route = (config: RouteConfig) => {
              const { path, component, name } = config;
              const Component = AllPages[component];

              return Component && (
                <Route
                  exact
                  key={index}
                  path={path}
                  render={(props: RouteComponentProps<any>) => {
                    // 匹配 location query 字段
                    const reg = /\?\S*/g;
                    const queryParams = window.location.hash.match(reg);
                    const { params } = props.match;
                    const title = `React Typescript Admin Starter Template-${name}`;

                    Object.keys(params).forEach((key: string) => {
                      params[key] = params[key] && params[key].replace(reg, '');
                    });

                    props.match.params = { ...params };

                    const mergeProps = {
                      app,
                      ...props,
                      query: queryParams ? queryString.parse(queryParams[0]) : {},
                    };

                    return this.requireLogin(
                      <DocumentTitle title={title}>
                        <Component {...mergeProps} />
                      </DocumentTitle>,
                      config.auth
                    );
                  }}
                />
              );
            };

            return config.component ? route(config) :
              config.children && config.children.map((d: RouteConfig) => route(d));
          })
        )}
        <Route path="/app/exception/403" component={NoPermission} />
        <Route path="/app/exception/404" component={NotFound} />
        <Route path="/app/exception/500" component={ServerError} />
        <Route render={() => <Redirect to="/app/exception/404" />} />
      </Switch>
    );
  }
}
