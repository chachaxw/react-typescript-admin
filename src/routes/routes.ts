// Global routes enum
export enum Routes {
  App = '/app',
  Login = '/login',

  // 首页
  Dashboard = '/app/dashboard',

  // 异常报错
  NoPermission = '/exception/403',
  NotFound = '/exception/404',
  ServerError = '/exception/500',
  AppNoPermission = '/app/exception/403',
  AppNotFound = '/app/exception/404',
  AppServerError = '/app/exception/500',
}

// 异常路由列表
export const ExceptionRoutes: string[] = [Routes.AppNoPermission, Routes.AppNotFound, Routes.AppServerError];
