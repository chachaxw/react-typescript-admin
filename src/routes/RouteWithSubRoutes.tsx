import { Route, RouteComponentProps } from 'dva/router';
import queryString from 'query-string';
import React, { FC } from 'react';
import DocumentTitle from 'react-document-title';

import { ServerEnv } from '../env';
import AllPages, { PublicPages } from '../pages';
import { RouteConfig } from './config';

const env = ServerEnv();
const { NotFound } = PublicPages;

interface Props extends RouteConfig {
  app: any;
}

export const RouteWithSubRoutes: FC<Props> = (route) => {
  const { path, name, app, children, auth, component } = route;
  const Component = component ? AllPages[component] : null;

  return (
    <Route
      path={path}
      render={(props: RouteComponentProps<any>) => {
        // Match location query string
        const reg = /\?\S*/g;
        const queryParams = window.location.href.match(reg);
        const { params } = props.match;
        const title = `${env.name !== '' ? env.name + '-' : ''} React Typescript Admin Starter Template-${name}`;

        Object.keys(params).forEach((key: string) => {
          params[key] = params[key] && params[key].replace(reg, '');
        });

        props.match.params = { ...params };

        const mergeProps = {
          app,
          auth,
          ...props,
          query: queryParams ? queryString.parse(queryParams[0]) : {},
        };

        return Component ? (
          <DocumentTitle title={title}>
            <Component {...mergeProps} routes={children} />
          </DocumentTitle>
        ) : (
          <NotFound />
        );
      }}
    />
  );
};

export default RouteWithSubRoutes;
