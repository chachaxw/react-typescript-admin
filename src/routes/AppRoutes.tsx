import { Redirect, Route, RouteComponentProps, Switch } from 'dva/router';
import queryString from 'query-string';
import React, { Component } from 'react';
import DocumentTitle from 'react-document-title';

import { ServerEnv } from '../env';
import { Auth } from '../models/typed';
import AllPages, { PublicPages } from '../pages';
import { UrlQueryReg } from '../utils/regTool';
import { RouteConfig } from './config';
import { Routes } from './routes';
import RouteWithSubRoutes from './RouteWithSubRoutes';

const env = ServerEnv();
const { NoPermission, NotFound, ServerError } = PublicPages;

interface InternalProps {
  app: any;
  auth: Auth;
  menu: RouteConfig[];
}

class AppRoutes extends Component<InternalProps> {
  private constructor(props: InternalProps) {
    super(props);
  }

  public requireAuth(permission: string, component: any) {
    const { auth } = this.props;
    const { permissions } = auth;

    if (!permissions || !permissions.includes(permission)) {
      return <Redirect to="/404" />;
    }

    return component;
  }

  public requireLogin(component: any, permission?: string) {
    const { auth } = this.props;
    const { isAuthenticated, permissions } = auth;

    if (!isAuthenticated || !permissions) {
      // 判断是否登录
      return <Redirect to={Routes.Login} key={Routes.Login} />;
    }

    return permission ? this.requireAuth(permission, component) : component;
  }

  public renderRoute(route: RouteConfig): any {
    if (route.component && route.children) {
      return this.requireLogin(<RouteWithSubRoutes app={this.props.app} {...route} />);
    }

    if (route.children) {
      return route.children.map((item) => this.renderRoute(item));
    }

    return this.renderComponent(route);
  }

  public renderComponent(config: RouteConfig) {
    const { app } = this.props;
    const { path, component, name } = config;
    const Component = component ? AllPages[component] : null;

    return (
      Component && (
        <Route
          exact
          key={path}
          path={path}
          render={(props: RouteComponentProps<any>) => {
            // 匹配 location query 字段
            const queryParams = window.location.hash.match(UrlQueryReg);
            const { params } = props.match;
            const title = `${env.name !== '' ? env.name + '-' : ''} React Typescript Admin Starter Template--${name}`;

            Object.keys(params).forEach((key: string) => {
              params[key] = params[key] && params[key].replace(UrlQueryReg, '');
            });

            props.match.params = { ...params };

            const mergeProps = {
              app,
              ...props,
              query: queryParams ? queryString.parse(queryParams[0]) : {},
            };

            return this.requireLogin(
              <DocumentTitle title={title}>
                <Component {...mergeProps} />
              </DocumentTitle>,
              config.auth
            );
          }}
        />
      )
    );
  }

  public render() {
    const { menu } = this.props;

    return (
      <Switch>
        {menu.map((config: RouteConfig) => this.renderRoute(config))}
        <Route path={Routes.AppNoPermission} component={NoPermission} />
        <Route path={Routes.AppNotFound} component={NotFound} />
        <Route path={Routes.AppServerError} component={ServerError} />
        <Route render={() => <Redirect to={Routes.AppNotFound} />} />
      </Switch>
    );
  }
}

export default AppRoutes;
