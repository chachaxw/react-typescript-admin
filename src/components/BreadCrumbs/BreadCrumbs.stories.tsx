import { storiesOf } from '@storybook/react';
import React from 'react';

import { appRoutes } from '../../routes';
import BreadCrumbs from './BreadCrumbs';

storiesOf('BreadCrumbs', module).add('default', () => <BreadCrumbs menu={appRoutes.app} url="/app/dashboard" />);
