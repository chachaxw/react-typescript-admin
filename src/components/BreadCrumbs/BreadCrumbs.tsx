import { Breadcrumb, Typography } from 'antd';
import { Link } from 'dva/router';
import React, { FC } from 'react';

import { RouteConfig } from '../../routes';
import { getBreadcrumbNameMap, urlToList } from '../../utils/utils';

const { Item } = Breadcrumb;
const { Text } = Typography;

// 渲染 Breadcrumb 子节点
// Render the Breadcrumb child node
interface DefaultItemProps {
  path: string;
  name: string;
  component?: string;
}

export const DefaultItem: FC<DefaultItemProps> = (props) => {
  const { path, component, name } = props;
  return component ? <Link to={path}>{name}</Link> : <Text>{name}</Text>;
};

interface Props {
  url: string;
  menu: RouteConfig[];
}

export const BreadCrumbs: FC<Props> = (props) => {
  const { url, menu } = props;
  const urlList = urlToList(url);
  const breadCrumbsMap = getBreadcrumbNameMap(menu);
  const breadCrumbs = urlList.map((item: string) => breadCrumbsMap[item]).filter((item) => item && item);

  if (!(breadCrumbs && breadCrumbs.length)) {
    return null;
  }

  return (
    <Breadcrumb style={{ margin: '12px 16px' }}>
      {breadCrumbs.map((item: RouteConfig) => (
        <Item key={item.path}>
          <DefaultItem path={item.path} name={item.name} component={item.component} />
        </Item>
      ))}
    </Breadcrumb>
  );
};

export default BreadCrumbs;
