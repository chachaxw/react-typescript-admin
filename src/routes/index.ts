import AppRoutes from './AppRoutes';
import { appRoutes, RouteConfig as RouteConfigTyped } from './config';
import { ExceptionRoutes, Routes } from './routes';
import RouteWithSubRoutes from './RouteWithSubRoutes';

export type RouteConfig = RouteConfigTyped;

export { AppRoutes, appRoutes, ExceptionRoutes, RouteWithSubRoutes, Routes };
