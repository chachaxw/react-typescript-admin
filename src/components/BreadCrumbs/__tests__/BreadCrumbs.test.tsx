import { render } from '@testing-library/react';
import { StaticRouter } from 'dva/router';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import { appRoutes, Routes } from '../../../routes';
import BreadCrumbs, { DefaultItem } from '../BreadCrumbs';

describe('DefaultItem test', () => {
  it('should display a text when without prop component', () => {
    const tree = shallow(<DefaultItem path={Routes.Dashboard} name="Dashboard" />);
    const props = tree.props();

    expect(props.children).toEqual('Dashboard');
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should display a link when with prop component', () => {
    const tree = shallow(
      <StaticRouter context={{}}>
        <DefaultItem path={Routes.Dashboard} name="Dashboard" component="Test" />
      </StaticRouter>
    );
    const { getByText } = render(
      <StaticRouter context={{}}>
        <DefaultItem path={Routes.Dashboard} name="Dashboard" component="Test" />
      </StaticRouter>
    );
    const link = getByText('Dashboard');

    expect(link.nodeName).toBe('A');
    expect(link.href).toBe('http://localhost/app/dashboard');
    expect(toJson(tree)).toMatchSnapshot();
  });
});

describe('BreadCrumbs test', () => {
  it('should display a breadcrumbs', () => {
    const tree = shallow(<BreadCrumbs menu={appRoutes.app} url={Routes.Dashboard} />);
    const props = tree.props();

    expect(props.children.length).toEqual(1);
    expect(toJson(tree)).toMatchSnapshot();
  });
});
