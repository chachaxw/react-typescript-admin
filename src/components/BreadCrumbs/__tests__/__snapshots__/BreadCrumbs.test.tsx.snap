// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadCrumbs test should display a breadcrumbs 1`] = `
<Breadcrumb
  style={
    Object {
      "margin": "12px 16px",
    }
  }
>
  <BreadcrumbItem
    key="/app/dashboard"
  >
    <DefaultItem
      component="Dashboard"
      name="控制台"
      path="/app/dashboard"
    />
  </BreadcrumbItem>
</Breadcrumb>
`;

exports[`DefaultItem test should display a link when with prop component 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <DefaultItem
    component="Test"
    name="Dashboard"
    path="/app/dashboard"
  />
</Router>
`;

exports[`DefaultItem test should display a text when without prop component 1`] = `
<Text>
  Dashboard
</Text>
`;
